### PR TITLE
Allwinner: add support for crust SCP firmware

### PR DESCRIPTION
**Pull request by dhewg**
Crust improves battery life and thermal performance by implementing a deep
sleep state. During deep sleep, the CPU cores, the DRAM controller, and most
onboard peripherals are powered down, reducing power consumption by 80% or more
compared to an idle device. On boards without a PMIC, Crust is also responsible
for orderly power-off and power-on of the device.

Depends on #4759

### CODE REVIEW

**Comment by dhewg** [comment]
added missing copyrights

**Comment by chewitt** [comment]
I'd normally object to bumping to an RC u-boot release but since Allwinner and Amlogic are both using the same 2021.01-rc3 base for current development and patches right now the practical thing to do is bump and then tidy up once 2021.01 ships mid-Jan. It might mean initial LE10 images ship with the RC code, but we're probably labelling things Alpha anyway so I'm not sure that matters.

**Comment by dhewg** [comment]
We can also wait for a release, but u-boot has a pretty good track record of stable RCs.
I don't mind either way, but you've also got a point with it not being the only pre-release package in master :)

**Comment by dhewg** [comment]
Updated crust to current master, which adds cir wakeup.
In its current state, it only supports the rc6 protocol, and the wakeup scancode is a build time config option.
It defaults to the mce power button though, which fits nicely for LE: [0x800f040c = "KEY_SLEEP"](https://git.linuxtv.org/v4l-utils.git/tree/utils/keytable/rc_keymaps/rc6_mce.toml?h=v4l-utils-1.20.0#n18)

File: packages/linux/package.mk
Comment (jernejsk) [inline packages/linux/package.mk:38]: Can you please move patches to Allwinner project folder? I would like to avoid polluting packages with project specifics. There are two ways this can be done without too much change:
1. put patches together with others in projects/Allwinner/patches/linux
2. put patches in projects/Allwinner/patches/linux/crust and set LINUX variable in project config to "crust", this is a bit hacky, but a way to have patches separated

File: packages/tools/u-boot/package.mk
Comment (jernejsk) [inline packages/tools/u-boot/package.mk:56]: Why are you changing ATF handling? I think this is out of scope for this PR. AFAIK it was made this way to support other projects when they start using open source replacements but I have no idea how far they are with that...

File: scripts/uboot_helper
Comment (jernejsk) [inline scripts/uboot_helper:105]: Please add other configs, at least those supported by Crust. If it works here, I'm pretty sure it will work with others too.

**Review by jernejsk — COMMENTED** [review]
Thanks for working on this! It was requested several times already.

**Comment by jernejsk** [comment]
@dhewg you can remove A83T related patches for now, they don't help anything...

File: packages/linux/package.mk
Comment (dhewg) [reply packages/linux/package.mk:38]: Moved the patches to projects/Allwinner/patches/linux/crust, turns out that already works with PKG_PATCH_DIRS

File: packages/tools/u-boot/package.mk
Comment (dhewg) [reply packages/tools/u-boot/package.mk:56]: It's a distinct patch, see https://github.com/LibreELEC/LibreELEC.tv/pull/4762/commits/2504672c37ab753d008e9a32cd908da93f20470d
We need some crust stuff in the u-boot package anyway (to express the dependency when required), so I figured we can move that one line too. Now it's all in one place, and that atf handling in not sunxi specific anyway.

File: scripts/uboot_helper
Comment (dhewg) [reply scripts/uboot_helper:113]: Added it for the boards where crust ships a defconfig

**Comment by dhewg** [comment]
> @dhewg you can remove A83T related patches for now, they don't help anything...

Done, removed the 3 a83 patches

File: packages/tools/u-boot/package.mk
Comment (jernejsk) [reply packages/tools/u-boot/package.mk:56]: @lrusak @CvH are you fine with this?

**Review by jernejsk — APPROVED** [review]

**Comment by jernejsk** [comment]
Thanks! I would like some input from others for changes in common infrastructure and packages (U-Boot) before merging.

**Comment by dhewg** [comment]
> Updated crust to current master, which adds cir wakeup.
> In its current state, it only supports the rc6 protocol, and the wakeup scancode is a build time config option.
> It defaults to the mce power button though, which fits nicely for LE: [0x800f040c = "KEY_SLEEP"](https://git.linuxtv.org/v4l-utils.git/tree/utils/keytable/rc_keymaps/rc6_mce.toml?h=v4l-utils-1.20.0#n18)

A little less technical, in case that wasn't clear: the power button of any "media center edition" compatible remote suspends the board when it's up and running, and wakes it up when it's suspending \o/

**Comment by dhewg** [comment]
Updated u-boot to v2021.01-rc4

**Comment by dhewg** [comment]
rc4 had no sunxi specific fixes, but there're a few for amlogic.
It sounds like there's going to be one more rc before the final release.

**Comment by CvH** [comment]
@dhewg needs rebase

**Comment by dhewg** [comment]
rebased

File: packages/tools/u-boot/package.mk
Comment (CvH) [inline packages/tools/u-boot/package.mk:35]: are you okay with that change ? @jernejsk @Kwiboo @chewitt

Comment (chewitt) [reply packages/tools/u-boot/package.mk:35]: I'm using the same so this is no drama for me. It should go final release mid-Jan.

Comment (jernejsk) [reply packages/tools/u-boot/package.mk:35]: I'm fine with that

File: packages/tools/u-boot/package.mk
Comment (Kwiboo) [reply packages/tools/u-boot/package.mk:56]: I am going to have to revert this change in #4591 since there is more files that needs to be copied on rockchip, see https://github.com/LibreELEC/LibreELEC.tv/pull/4591/files#diff-1cdbf9d1405af32af934d0e88882b5449c0fdc93a55b7f646f7f21583185ff1e

Comment (dhewg) [reply packages/tools/u-boot/package.mk:56]: Oh alright, I wasn't aware that RK requires additional files

**Pull request merged**
